libs/pruviloj/Pruviloj.idr

```diff
@@ -0,0 +1,115 @@
+module Pruviloj
+
+import Language.Reflection.Elab
+import Language.Reflection.Utils
+
+data Infer : Type where
+  MkInfer : (a : Type) -> a -> Infer
+
+||| Run something for effects, throwing away the return value
+ignore : Functor f => f a -> f ()
+ignore x = map (const ()) x
+
+||| Do nothing
+skip : Applicative f => f ()
+skip = pure ()
+
+||| Attempt to apply a tactic. If it fails, do nothing.
+try : Elab a -> Elab ()
+try tac = tac *> pure () <|> pure ()
+
+||| Solve the goal using the most recent applicable hypothesis
+hypothesis : Elab ()
+hypothesis =
+    do hyps <- map fst <$> getEnv
+       choiceMap (\n => do fill (Var n); solve)
+                 hyps
```

Comment:
This seems like it could look nicer with `flip choiceMap`, i.e. something like

```idris
       flip choiceMap hyps $ \n => do
         fill (Var n)
         solve
```

If this is a frequent idiom, a dedicated combinator may be appropriate, as with `traverse`/`for`.